api_server.go

```diff
@@ -0,0 +1,196 @@
+		if err != nil {
+			logger.WithError(err).Error("template.ParseFiles() failed")
+			return nil, http.StatusInternalServerError
+		}
+		if err := musicPDFsTemplate.Execute(&buf, &pdfs); err != nil {
+			logger.WithError(err).Error("template.Execute() failed")
+			return nil, http.StatusInternalServerError
+		}
+	default:
+		if err := json.NewEncoder(&buf).Encode(&objects); err != nil {
+			logger.WithError(err).Error("json.Encode() failed")
+			return nil, http.StatusInternalServerError
+		}
+	}
+	return buf.Bytes(), http.StatusOK
+}
+
+func acceptsType(r *http.Request, mimeType string) bool {
+	for _, t := range strings.Split(r.Header.Get("Accept"), ",") {
+		if t == mimeType {
+			return true
+		}
+	}
+	return false
+}
+
+func (x *ApiServer) MusicPDFsUpload(r *http.Request) ([]byte, int) {
+	// only accept post
+	if r.Method != http.MethodPost {
+		return nil, http.StatusMethodNotAllowed
+	}
+
+	// only allow <1MB
+	if r.ContentLength > int64(1e6) {
+		return nil, http.StatusRequestEntityTooLarge
+	}
+
+	// read the metadata
+	meta := NewMusicPDFMetaFromUrlValues(r.URL.Query())
+	if err := meta.Validate(); err != nil {
+		return []byte(err.Error()), http.StatusBadRequest
+	}
+
+	// read the pdf from the body
+	var pdfBytes bytes.Buffer
+	if _, err := pdfBytes.ReadFrom(r.Body); err != nil {
+		logger.WithError(err).Println("r.body.Read() failed")
+		return nil, http.StatusBadRequest
+	}
+
+	// write the pdf
+	object := Object{
+		ContentType: "application/pdf",
+		Name:        fmt.Sprintf("%s-%s-%d.pdf", meta.Project, meta.Instrument, meta.PartNumber),
+		Tags:        meta.ToTags(),
+		Buffer:      pdfBytes,
+	}
+	if err := x.PutObject(MusicPdfsBucketName, &object); err != nil {
+		logger.WithError(err).Println("storage.PutObject() failed")
+		return nil, http.StatusInternalServerError
+	}
+	return nil, http.StatusOK
+}
+
+const MusicPdfsBucketName = "music-pdfs"
+
+var (
+	ErrMissingProject    = fmt.Errorf("missing required field `project`")
+	ErrMissingInstrument = fmt.Errorf("missing required field `instrument`")
+	ErrMissingPartNumber = fmt.Errorf("missing required field `part_number`")
+)
+
+type MusicPDFMeta struct {
+	Project      string
+	Instrument   string
+	PartNumber   int
+	DownloadLink string
+}
+
+func NewMusicPDFMetaFromTags(tags Tags) MusicPDFMeta {
+	partNumber, _ := strconv.Atoi(tags["Part-Number"])
+	return MusicPDFMeta{
+		Project:    tags["Project"],
+		Instrument: tags["Instrument"],
+		PartNumber: partNumber,
+	}
+}
+
+func NewMusicPDFMetaFromUrlValues(values url.Values) MusicPDFMeta {
+	partNumber, _ := strconv.Atoi(values.Get("part_number"))
+	return MusicPDFMeta{
+		Project:    values.Get("project"),
+		Instrument: values.Get("instrument"),
+		PartNumber: partNumber,
+	}
+}
+
+func (x *MusicPDFMeta) ToTags() map[string]string {
+	return map[string]string{
+		"Project":     x.Project,
+		"Instrument":  x.Instrument,
+		"Part-Number": strconv.Itoa(x.PartNumber),
+	}
+}
```

Comment:
What happens on conversion error?

Reply:
There's no conversion error that can happen here. All the types can be converted without fail.